src/main/scala/com/fulcrumgenomics/util/AssignPrimers.scala

```diff
@@ -0,0 +1,245 @@
+ @arg(doc="The SAM tag for the mate's assigned amplicon identifier.") val mateAmpliconIdentifierTag: String = AssignPrimers.MateAmpliconIdentifierTag,
+) extends FgBioTool with LazyLogging {
+
+  Io.assertReadable(input)
+  Io.assertReadable(primers)
+  Io.assertCanWriteFile(output)
+
+  override def execute(): Unit = {
+    val reader    = SamSource(input)
+    val writer    = SamWriter(output, reader.header)
+    val progress  = ProgressLogger(logger=logger, unit=250000)
+    val amplicons = Metric.read[Amplicon](path=primers)
+    val detector  = new AmpliconDetector(
+      detector             = Amplicon.detector(amplicons=amplicons.iterator),
+      slop                 = slop,
+      unclippedCoordinates = unclippedCoordinates
+    )
+
+    val metricsMap: Map[Amplicon, AssignPrimersMetric] = amplicons.map { amplicon =>
+      amplicon -> new AssignPrimersMetric(identifier=amplicon.identifier)
+    }.toMap
+
+    val labeller = new AmpliconLabeller(
+      metrics                   = metricsMap,
+      primerCoordinatesTag      = primerCoordinatesTag,
+      matePrimerCoordinatesTag  = matePrimerCoordinatesTag,
+      ampliconIdentifierTag     = ampliconIdentifierTag,
+      mateAmpliconIdentifierTag = mateAmpliconIdentifierTag
+    )
+
+    reader.foreach { rec =>
+      val recAmplicon = detector.find(rec=rec)
+      labeller.label(
+        rec = rec,
+        recAmplicon = recAmplicon,
+        mateAmplicon = if (rec.unpaired) None else detector.findMate(rec=rec)
+      )
+      writer.write(rec)
+      progress.record(rec)
+    }
+    progress.logLast()
+
+    reader.safelyClose()
+    writer.close()
+
+    // Sum up the values across all amplicons
+    val totalMetric = new AssignPrimersMetric(identifier=AssignPrimersMetric.AllAmpliconsIdentifier)
+    metricsMap.values.foreach { metric =>
+      totalMetric.r1s   += metric.r1s
+      totalMetric.r2s   += metric.r2s
+      totalMetric.left  += metric.left
+      totalMetric.right += metric.right
+      totalMetric.pairs += metric.pairs
+    }
+
+    // Write the metrics
+    Metric.write[AssignPrimersMetric](
+      path    = metrics,
+      metrics = (amplicons.iterator.map(metricsMap.apply) ++ Iterator(totalMetric)).map(_.finalize(total=progress.getCount))
+    )
+
+    // Log some info
+    def log(numerator: Long, noun: String): Unit = {
+      val pct: Double = if (progress.getCount == 0) 0 else numerator * 100.0 / progress.getCount
+      logger.info(f"Assigned $numerator%,d out of ${progress.getCount}%,d ($pct%.2f%%) reads $noun.")
+    }
+    log(numerator=totalMetric.left, "to left primers")
+    log(numerator=totalMetric.right, "to right primers")
+    log(numerator=totalMetric.pairs * 2, "as primer pairs")
+  }
+}
+
+object AssignPrimers {
+  /** The SAM tag to use for the current record's assigned primer genomic coordinates */
+  val PrimerCoordinateTag      : String = "ap"
```

Comment:
Just a note on the tag keys.  It's weird to me that the "primer" ones start with `a` and the amplicons ones don't!  I'd suggest `rp` and `mp` for read-primer and mate-primer and `ra` and `ma` for read amplicon and mate amplicon.